### PR TITLE
[SYCL][NVPTX] Split max_work_group_size into 3 NVVM annotations

### PR DESCRIPTION
**Pull request by frasercrmck**
NVVM IR supports separated maxntidx, maxntidy, and maxntidz annotations. The backend will print them individually as three dimensions. This better preserves programmer intent than prematurely flattening them together.

Note that the semantics are in fact identical; the CUDA implementation internally multiplies all dimensions together and only guarantees that the total is never exceeded, but not that any individual dimension is not exceeded. Thus 64,1,1 is identical to 4,4,4.

We try and preserve a logical mapping of dimensions by index flipping between SYCL (z,y,x) and NVVM (x,y,z) in CUDA terminology despite, as mentioned above, it being largely irrelevant.

Also this patch simplifies the attribute's getter functions as all dimensions are mandatory, and the getters seemed copied from the reqd_work_group_size attribute where some are optional.

We could probably improve the code further by making the operands "unsigned" and not "Expr", and renaming them from X,Y,Z to Dim{0,1,2} as per the SYCL spec. This has been left for future work, however, as there's a non-trivial amount of code that expects to be able to treat the max_work_group_size and reqd_work_group_size attributes identically through templates and identical helper methods.

### CODE REVIEW

**Review by steffenlarsen — APPROVED** [review]
Great improvement! 🚀

File: clang/include/clang/Basic/Attr.td [hunk 1]
```diff
@@ -1705,20 +1705,17 @@ def SYCLIntelMaxWorkGroupSize : InheritableAttr {
   let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
   let Subjects = SubjectList<[Function], ErrorDiag>;
   let AdditionalMembers = [{
-    std::optional<llvm::APSInt> getXDimVal() const {
-      if (const auto *CE = dyn_cast<ConstantExpr>(getXDim()))
-        return CE->getResultAsAPSInt();
-      return std::nullopt;
+    unsigned getXDimVal() const {
+      assert(isa<ConstantExpr>(getXDim()));
```
Comment (jchlanda) [inline clang/include/clang/Basic/Attr.td:1709]: NIT: I don't think you need to assert on `isa<T>` here, [`cast<T>` will do it](https://github.com/jchlanda/llvm/blob/sycl/llvm/include/llvm/Support/Casting.h#L577) for you (here and in `Y`, `Z`).

**Review by jchlanda — APPROVED** [review]
🏅

File: clang/include/clang/Basic/Attr.td (hunk 1 above)
Comment (frasercrmck) [reply clang/include/clang/Basic/Attr.td:1709]: Very true, in fact, `you should not use an isa<> test followed by a cast<>, for that use the dyn_cast<> operator.`[1] - though I see it used in (mostly older) LLVM code. If it gives a more meaningful assertion message to the user than the `cast` assertion I could maybe see this pattern's utility It's also not clear from the guidance whether this is refers to conditional cases, like `if (isa<X>(V)) { cast<X>(V); }` or whether it means ever, like ever ever.

As it stands I haven't provided a useful assertion message so you're right to call it out, thanks!

Looking around similar attribute-handling code in clang, I suspect we are doing something slightly off as it's only really SYCL attributes that have casts to "ConstantExpr". I wonder if this is due to us wanting to use attributes with templated values? I'd have to look into it deeper. As it stands, enough SYCL attributes just blindly `cast` so I'll do it here too, for consistency.

[1] https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates

Comment (frasercrmck) [reply clang/include/clang/Basic/Attr.td]: done!

**Comment by frasercrmck** [comment]
@intel/llvm-gatekeepers this is good to merge, thank you!

**Pull request merged**